File: 1402-reducing-dishes/1402-reducing-dishes.js
/**
 * @param {number[]} satisfaction
 * @return {number}
 */

let getMaxSatisfaction = (satisfaction, index, time, memo={}) => {
    let key = index + '-' + time
    
    if(key in memo)
        return memo[key]
    if(index >= satisfaction.length)
        return 0
    
    // either we include a dish or exclude it
    
    // if we include it
    let includeSatisfaction = (satisfaction[index] * time) + getMaxSatisfaction(satisfaction, index+1, time+1, memo)
    
    // if we do not include it
    let excludeSatisfaction = getMaxSatisfaction(satisfaction, index+1, time, memo)
    
    memo[key] = Math.max(includeSatisfaction, excludeSatisfaction)
    
    return memo[key]
}


var maxSatisfaction = function(satisfaction) {
    // sort the satisfaction array
    satisfaction = satisfaction.sort((a, b) => a-b)
        
    // call getMaxSatisfaction
    return getMaxSatisfaction(satisfaction, 0, 1)
};

File: 1402-reducing-dishes/1402-reducing-dishes.java
class Solution {
    int[][] memo;
    public int maxSatisfaction(int[] satisfaction) {
        // for every dish either take or dont take
        
        // sort the array in ascending order
        Arrays.sort(satisfaction);
        int n = satisfaction.length;
        
        memo = new int[n+1][n+1];
        for(int i=0; i<=n; i++) {
            for(int j=0; j<=n; j++) {
                memo[i][j] = -1;
            }
        }
        return getMaxSatisfaction(satisfaction, 0, 1);
    }
    
    public int getMaxSatisfaction(int[] satisfaction, int dish, int time) {
        if(dish == satisfaction.length)
            return 0;
        
        if(memo[dish][time] != -1)
            return memo[dish][time];
        
        int currentSatisfaction = time * satisfaction[dish];
        
        memo[dish][time] = Math.max(currentSatisfaction+ getMaxSatisfaction(satisfaction, dish+1, time+1), getMaxSatisfaction(satisfaction, dish+1, time));
        
        return memo[dish][time];
    }
}